node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/file/FileHelper.java

```diff
@@ -0,0 +1,263 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.node.admin.task.util.file;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Stack;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import static com.yahoo.vespa.hosted.node.admin.task.util.file.IOExceptionUtil.uncheck;
+
+/**
+ * @author freva
+ */
+public class FileHelper {
+
+    /**
+     * Creates a {@link FileStreamBuilder} that will by default match all files and all directories
+     * under the given basePath.
+     */
+    public FileStreamBuilder streamContents(Path basePath) {
+        return new FileStreamBuilder(basePath, true, true);
+    }
+
+    /**
+     * Creates a {@link FileStreamBuilder} that will by default match all files and no directories
+     * under the given basePath.
+     */
+    public FileStreamBuilder streamFiles(Path basePath) {
+        return new FileStreamBuilder(basePath, true, false);
+    }
+
+    /**
+     * Creates a {@link FileStreamBuilder} that will by default match all directories and no files
+     * under the given basePath.
+     */
+    public FileStreamBuilder streamDirectories(Path basePath) {
```

Comment:
FileHelper is an unusual class:
 - factory methods like streamDirectories() is non-static
 - builder class is non-static
 - builder class calls method of enclosing class

I'm not entirely convinced this is a good design. For instance the factory method could be static, and the entire body of FileStreamBuilder could be moved to FileHelper, and the FileStreamBuilder inner class can be removed?

Reply:
Fixed